src/hotspot/share/prims/universalUpcallHandler.cpp

```diff
@@ -36,6 +36,26 @@
 
 extern struct JavaVM_ main_vm;
 
+// When an upcall is invoked from a thread that is not attached to the VM, we need to attach it,
+// and then to detach it at some point later. Detaching a thread as soon as the upcall completes
+// is suboptimal, as the same thread could later upcall to Java again, at which point the VM would
+// create multiple Java views of the same native thread. For this reason, we use thread local storage
+// to keep track of the fact that we have attached a native thread to the VM. When the thread local
+// storage is destroyed (which happens when the native threads is terminated), we check if the
+// storage has an attached thread and, if so, we detach it from the VM.
+struct UpcallContext {
+  Thread* attachedThread;
+
+  ~UpcallContext() {
+    if (attachedThread != NULL) {
+      JavaVM_ *vm = (JavaVM *)(&main_vm);
+      vm->functions->DetachCurrentThread(vm);
+    }
+  }
+};
+
+thread_local UpcallContext threadContext;
```

Comment:
Use of C++11 `thread_local` violates the Hotspot rules for C++ feature usage. See:
https://github.com/openjdk/jdk/blob/master/doc/hotspot-style.md#thread_local
It is also unclear when such an object will be destroyed when C++ thread API is not used.